src/procedures/proc_spmw_paths.c

```diff
@@ -0,0 +1,471 @@
+				if(allPathsCtx->dst != NULL) {
+					Node dst = Path_Head(allPathsCtx->path);
+					if(ENTITY_GET_ID(allPathsCtx->dst) != ENTITY_GET_ID(&dst)) continue;
+				}
+				p->path = allPathsCtx->path;
+				return;
+			}
+		} else {
+			// No way to advance, backtrack.
+			Path_PopNode(allPathsCtx->path);
+			if(Path_EdgeCount(allPathsCtx->path)) {
+				Edge e = Path_PopEdge(allPathsCtx->path);
+				SIValue c = _get_value_or_defualt((GraphEntity *)&e, ctx->cost_prop, SI_LongVal(1));
+				SIValue w = _get_value_or_defualt((GraphEntity *)&e, ctx->weight_prop, SI_LongVal(1));
+				p->cost -= SI_GET_NUMERIC(c);
+				p->weight -= SI_GET_NUMERIC(w);
+			}
+		}
+	}
+
+	// Couldn't find a path.
+	p->path = NULL;
+	return;
+}
+
+static int path_cmp(const void *a, const void *b, const void *udata) {
+	WeightedPath *da = (WeightedPath *)a;
+	WeightedPath *db = (WeightedPath *)b;
+	if(da->weight == db->weight) {
+		if(da->cost - db->cost) {
+			return Path_Len(da->path) - Path_Len(db->path);
+		}
+		return da->cost - db->cost;
+	}
+	return da->weight - db->weight;
+}
+
+static void SPMWpaths_all_minimal
+(
+	SPMWctx *spmw_ctx
+) {
+	spmw_ctx->array = array_new(WeightedPath, 0);
+
+	WeightedPath p = {0};
+	double max_weight = DBL_MAX;
+	SPMWpaths_next(spmw_ctx, &p, max_weight);
+	while (p.path != NULL) {
+		uint count = array_len(spmw_ctx->array);
+		if(count > 0 && p.weight < spmw_ctx->array[0].weight) {
+			for (uint i = 0; i < array_len(spmw_ctx->array); i++) {
+				Path_Free(spmw_ctx->array[i].path);
+			}
+			array_clear(spmw_ctx->array);
+			count = 0;
+		}
+		p.path = Path_Clone(p.path);
+		array_append(spmw_ctx->array, p);
+		max_weight = p.weight;
+		SPMWpaths_next(spmw_ctx, &p, max_weight);
+	}
+}
+
+static void SPMWpaths_single_minimal
+(
+	SPMWctx *spmw_ctx
+) {
+	spmw_ctx->single.path   = NULL;
+	spmw_ctx->single.weight = DBL_MAX;
+	spmw_ctx->single.cost   = DBL_MAX;
+	WeightedPath p = {0};
+	SPMWpaths_next(spmw_ctx, &p, DBL_MAX);
+	while (p.path != NULL) {
+		if(p.weight < spmw_ctx->single.weight ||
+			(p.weight == spmw_ctx->single.weight &&
+				(p.cost < spmw_ctx->single.cost || Path_Len(p.path) < Path_Len(spmw_ctx->single.path)))) {
+			spmw_ctx->single.path = Path_Clone(p.path);
+			spmw_ctx->single.weight = p.weight;
+			spmw_ctx->single.cost = p.cost;
+		}
+		SPMWpaths_next(spmw_ctx, &p, spmw_ctx->single.weight);
+	}
+}
+
+static void SPMWpaths_k_minimal
+(
+	SPMWctx *spmw_ctx
+) {
+	spmw_ctx->heap = Heap_new(path_cmp, NULL);
+
+	WeightedPath p = {0};
+	double max_weight = DBL_MAX;
+	SPMWpaths_next(spmw_ctx, &p, max_weight);
+	while (p.path != NULL) {
+		if(Heap_count(spmw_ctx->heap) == spmw_ctx->path_count) {
+			WeightedPath *pp = Heap_peek(spmw_ctx->heap);
+			if(p.weight < pp->weight) {
+				Heap_poll(spmw_ctx->heap);
+				Path_Free(pp->path);
+				pp->path = Path_Clone(p.path);
+				pp->weight = p.weight;
+				pp->cost = p.cost;
+				Heap_offer(&spmw_ctx->heap, pp);
+			}
+			max_weight = pp->weight;
```

Comment:
to determine the new max weight, we need to peek at the heap and get the wight of the worse candidate among the best paths we've accepted so far.